### ISSUE TITLE
IllegalStateException "AssetManager has been finalized!"

### ISSUE THREAD

**Issue by inio** [google-ar/arcore-android-sdk#563]
We've seen a couple reports of this issue.

In some situations, an ARCore operation will fail with a FatalException and the logcat will show:

    java.lang.IllegalStateException: AssetManager has been finalized!
    ...
    android.content.ContextWrapper.createPackageContext(ContextWrapper.java:851)

We are still searching for the root cause of this error.

**Comment by inio** [comment]
Replying for bookkeeping purposes.

**Comment by inio** [comment]
@wuzeqin and @Calence have seen this.

**Comment by inio** [comment]
@siny-yao sees this show up as a crash in native code in session create, with the actual failure in `art::ScopedCheck::CheckThread(_JNIEnv*)` due to a pending exception (#584).  @wuzeqin separately sees the `art::ScopedCheck::CheckThread(_JNIEnv*)` crash with an identical stack, which I'm going to to assume is caused by the same pending exception.

**Comment by siny-yao** [comment]
> which I'm going to to assume is caused by the same pending exception.

This seems to be the same problem

**Comment by inio** [comment]
In previous attempts we'd been unable to reproduce this issue. To help test a workaround, could someone who is able to reliably reproduce this issue try the following?

In your `Activity`, add the following bits:

at the class scope:

```
static Context remoteArCoreContext;
static AssetManager remoteArCoreAssets;
```

in `onCreate`:

```
if (remoteArCoreContext == null) {
  remoteArCoreContext = context.createPackageContext("com.google.ar.core", 0);
  remoteArCoreAssets = remoteArCoreContext.getAssets();
}
```

Our current best theory is that this is actually an Android bug where the remote asset loader (that we use to load stuff from ARCore APK, like device calibration) is getting garbage collected and then can't be recreated. Holding a copy of it would prevent the garbage collection from happening.

**Comment by inio** [comment]
@wuzeqin, @Calence, @siny-yao: If any of you can reliably produce the "AssetManager has been finalized" error, please try adding the above lines to your Activity and let me know if you are still able to produce the failure.

**Comment by wuzeqin** [comment]
I try your advice in SM-G9600/DS the "AssetManager has been finalized" error still occur .
Wacky thing is that once the crash occur,  app reopen ,   ArSession_create method must crash. 
But I found that when crash occurred ,  the only way to solve it was uninstalling the ARCore.apk and reloading a new one.

**Comment by inio** [comment]
Thanks for trying that.  Good news is we've found a way of reproducing this reliably and are looking into fixes.

**Comment by wuzeqin** [comment]
May I ask you about the progress of fixing this bug ？ Wait for your good news. Thank you.

**Comment by inio** [comment]
None yet, but haven't had cycles to look at this recently.  Hoping to have some time before 1.6 branches.

**Comment by inio** [comment]
We've just released an update (1.5.1) with a minor improvement on this:  The crash some users started seeing in 1.5 should now be back to a FatalException during session create.  This update is available through the Play Store and can be downloaded on our [releases page](https://github.com/google-ar/arcore-android-sdk/releases).

**Comment by fredsa** [comment]
Fixed in ARCore for Unity SDK 1.6.0
